{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Coevolution Streptococcus-Lactobacilus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtain data from servers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, sys\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "264199\n",
      "Streptococcus thermophilus (strain ATCC BAA-250 / LMG 18311)\n",
      "299768\n",
      "Streptococcus thermophilus (strain CNRZ 1066)\n",
      "322159\n",
      "Streptococcus thermophilus (strain ATCC BAA-491 / LMD-9)\n",
      "767463\n",
      "Streptococcus thermophilus (strain ND03)\n",
      "1042404\n",
      "Streptococcus thermophilus CNCM I-1630\n",
      "1051074\n",
      "Streptococcus thermophilus JIM 8232\n",
      "1073569\n",
      "Streptococcus thermophilus MTCC 5460\n",
      "1073570\n",
      "Streptococcus thermophilus MTCC 5461\n",
      "1091038\n",
      "Streptococcus thermophilus DSM 20617\n",
      "1187956\n",
      "Streptococcus thermophilus MN-ZLW-002\n",
      "1263110\n",
      "Streptococcus thermophilus CAG:236\n",
      "1268061\n",
      "Streptococcus thermophilus DGCC 7710\n",
      "1408178\n",
      "Streptococcus thermophilus ASCC 1275\n",
      "1415776\n",
      "Streptococcus thermophilus TH1435\n",
      "1423145\n",
      "Streptococcus thermophilus TH1436\n",
      "1433288\n",
      "Streptococcus thermophilus MTH17CL396\n",
      "1433289\n",
      "Streptococcus thermophilus M17PTZA496\n",
      "1435972\n",
      "Streptococcus thermophilus TH985\n",
      "1435974\n",
      "Streptococcus thermophilus TH982\n",
      "1435981\n",
      "Streptococcus thermophilus 1F8CT\n",
      "1436725\n",
      "Streptococcus thermophilus TH1477\n",
      "1302\n",
      "Streptococcus gordonii\n",
      "1303\n",
      "Streptococcus oralis\n",
      "1304\n",
      "Streptococcus salivarius\n",
      "1305\n",
      "Streptococcus sanguinis\n",
      "1306\n",
      "Streptococcus sp.\n",
      "1307\n",
      "Streptococcus suis\n",
      "1308\n",
      "Streptococcus thermophilus\n",
      "1309\n",
      "Streptococcus mutans\n",
      "1310\n",
      "Streptococcus sobrinus\n",
      "1311\n",
      "Streptococcus agalactiae\n",
      "1313\n",
      "Streptococcus pneumoniae\n",
      "1314\n",
      "Streptococcus pyogenes\n",
      "1317\n",
      "Streptococcus downei\n",
      "1318\n",
      "Streptococcus parasanguinis\n",
      "1319\n",
      "Streptococcus sp. 'group B'\n",
      "1320\n",
      "Streptococcus sp. group G\n",
      "1324\n",
      "Streptococcus sp. G148\n",
      "1325\n",
      "Streptococcus sp. GX7805\n",
      "1326\n",
      "Streptococcus acidominimus\n",
      "1328\n",
      "Streptococcus anginosus\n",
      "1329\n",
      "Streptococcus canis\n",
      "1332\n",
      "Streptococcus criae\n",
      "1333\n",
      "Streptococcus criceti\n",
      "1334\n",
      "Streptococcus dysgalactiae\n",
      "1335\n",
      "Streptococcus equinus\n",
      "1336\n",
      "Streptococcus equi\n",
      "1337\n",
      "Streptococcus hyointestinalis\n",
      "1338\n",
      "Streptococcus intermedius\n",
      "1339\n",
      "Streptococcus macacae\n",
      "1340\n",
      "Streptococcus porcinus\n",
      "1341\n",
      "Streptococcus ratti\n",
      "1343\n",
      "Streptococcus vestibularis\n",
      "1345\n",
      "Streptococcus ferus\n",
      "1346\n",
      "Streptococcus iniae\n",
      "1348\n",
      "Streptococcus parauberis\n",
      "1349\n",
      "Streptococcus uberis\n",
      "10728\n",
      "Streptococcus pneumoniae phage HB-3\n",
      "10747\n",
      "Streptococcus phage Cp-1\n",
      "10748\n",
      "Streptococcus phage Cp-7\n",
      "10749\n",
      "Streptococcus phage Cp-9\n",
      "12344\n",
      "Streptococcus phage Cp-5\n",
      "12366\n",
      "Streptococcus pyogenes phage H4489A\n",
      "12402\n",
      "Streptococcus phage EJ-1\n",
      "28037\n",
      "Streptococcus mitis\n",
      "29389\n",
      "Streptococcus alactolyticus\n",
      "29390\n",
      "Streptococcus gordonii str. Challis\n",
      "33040\n",
      "Streptococcus milleri\n",
      "33972\n",
      "Streptococcus sp. 'group C'\n",
      "35344\n",
      "Streptococcus pyogenes phage T12\n",
      "36470\n",
      "Streptococcus sp. 'group A'\n",
      "39425\n",
      "Streptococcus phage T270\n",
      "40041\n",
      "Streptococcus equi subsp. zooepidemicus\n",
      "40287\n",
      "Streptococcus mutans serotype C\n",
      "45634\n",
      "Streptococcus cristatus\n",
      "53354\n",
      "Streptococcus gallolyticus\n",
      "55085\n",
      "Streptococcus thoraltensis\n",
      "59310\n",
      "Streptococcus macedonicus\n",
      "64186\n",
      "Streptococcus virus Sfi21\n",
      "68891\n",
      "Streptococcus peroris\n",
      "68892\n",
      "Streptococcus infantis\n",
      "69017\n",
      "Streptococcus sp. (strain 19909)\n",
      "72638\n",
      "Streptococcus virus Sfi19\n",
      "73422\n",
      "Streptococcus phage TP-J34\n",
      "73492\n",
      "Streptococcus pyogenes phage\n",
      "74382\n",
      "Streptococcus phage SFi18\n",
      "76860\n",
      "Streptococcus constellatus\n",
      "78535\n",
      "Streptococcus viridans\n",
      "78541\n",
      "Streptococcus virus Sfi11\n",
      "82269\n",
      "Streptococcus sp. 28D\n",
      "82348\n",
      "Streptococcus pluranimalium\n",
      "82806\n",
      "Streptococcus ovis\n",
      "83545\n",
      "Streptococcus sp. KN1\n",
      "83546\n",
      "Streptococcus sp. KN2\n",
      "83547\n",
      "Streptococcus sp. KN3\n",
      "83549\n",
      "Streptococcus sp. TW1\n",
      "85154\n",
      "Streptococcus phage O1205\n",
      "86065\n",
      "Streptococcus pyogenes phage H10403\n",
      "90410\n",
      "Streptococcus virus DT1\n",
      "99822\n",
      "Streptococcus dysgalactiae subsp. dysgalactiae\n",
      "102143\n",
      "Streptococcus phage S3b\n",
      "102144\n",
      "Streptococcus phage S92\n",
      "102145\n",
      "Streptococcus phage ST3\n",
      "102146\n",
      "Streptococcus phage ST64\n",
      "102147\n",
      "Streptococcus phage Sfi16A\n",
      "102150\n",
      "Streptococcus phage J1\n",
      "102684\n",
      "Streptococcus infantarius\n",
      "102886\n",
      "Streptococcus didelphis\n",
      "104215\n",
      "Streptococcus sp. 1400-98\n",
      "112023\n",
      "Streptococcus virus 7201\n",
      "113107\n",
      "Streptococcus australis\n",
      "114652\n",
      "Streptococcus orisratti\n",
      "116154\n",
      "Streptococcus sp. Z1227\n",
      "116155\n",
      "Streptococcus sp. Z12\n",
      "116156\n",
      "Streptococcus sp. Z89\n",
      "118670\n",
      "Streptococcus sp. PSH2\n",
      "118671\n",
      "Streptococcus sp. PSH1a\n",
      "118672\n",
      "Streptococcus sp. PSH1b\n",
      "119224\n",
      "Streptococcus phocae\n",
      "119602\n",
      "Streptococcus dysgalactiae subsp. equisimilis\n",
      "119603\n",
      "Streptococcus dysgalactiae group\n"
     ]
    }
   ],
   "source": [
    "def load_taxa(scientific_prefix):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    requestURL = \"https://www.ebi.ac.uk/proteins/api/taxonomy/name/\" + scientific_prefix +\\\n",
    "                \"%20?pageNumber=1&pageSize=100&searchType=STARTSWITH&fieldName=SCIENTIFICNAME\"\n",
    "\n",
    "    r = requests.get(requestURL, headers={ \"Accept\" : \"application/json\"})\n",
    "\n",
    "    if not r.ok:\n",
    "      r.raise_for_status()\n",
    "      sys.exit()\n",
    "\n",
    "    jsonBody = json.loads(r.text)\n",
    "    taxa = []\n",
    "    names = []\n",
    "    for taxonomy in jsonBody[\"taxonomies\"]:\n",
    "        print(taxonomy['taxonomyId'])\n",
    "        print(taxonomy['scientificName'])\n",
    "        taxa.append(taxonomy['taxonomyId'])\n",
    "        names.append(taxonomy['scientificName'])\n",
    "    return taxa, names\n",
    "\n",
    "termophilus_taxa,  termophilus_names = load_taxa(\"Streptococcus thermophilus\")\n",
    "streptococcus_taxa,  streptococcus_names = load_taxa(\"Streptococcus\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1302, 1303, 1304, 1305, 1306, 1307, 1308, 1309, 1310, 1311, 1313, 1314, 1317, 1318, 1319, 1320, 1324, 1325, 1326]\n",
      "[264199, 299768, 322159, 767463, 1042404, 1051074, 1073569, 1073570, 1091038, 1187956, 1263110, 1268061, 1408178, 1415776, 1423145, 1433288, 1433289, 1435972, 1435974]\n",
      "1302,1303,1304,1305,1306,1307,1308,1309,1310,1311,1313,1314,1317,1318,1319,1320,1324,1325,1326\n",
      "https://www.ebi.ac.uk/proteins/api/proteins?offset=0&size=-1&taxid=1302,1303,1304,1305,1306,1307,1308,1309,1310,1311,1313,1314,1317,1318,1319,1320,1324,1325,1326&reviewed=false&gene=LDH,CAS2,CAS3\n",
      "264199,299768,322159,767463,1042404,1051074,1073569,1073570,1091038,1187956,1263110,1268061,1408178,1415776,1423145,1433288,1433289,1435972,1435974\n",
      "https://www.ebi.ac.uk/proteins/api/proteins?offset=0&size=-1&taxid=264199,299768,322159,767463,1042404,1051074,1073569,1073570,1091038,1187956,1263110,1268061,1408178,1415776,1423145,1433288,1433289,1435972,1435974&reviewed=false&gene=LDH,CAS2,CAS3\n"
     ]
    }
   ],
   "source": [
    "def load_proteome(taxids, size=10, protein=[\"LDH\"]):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    taxids_str = \",\".join(str(x) for x in taxids)\n",
    "    protein_str = \",\".join(x for x in protein)\n",
    "    print(taxids_str)\n",
    "    requestURL = \"https://www.ebi.ac.uk/proteins/api/proteins?offset=0&size=\" + str(size) + \"&taxid=\" +\\\n",
    "                    taxids_str + \"&reviewed=false\"\n",
    "    if protein != []:\n",
    "        requestURL += \"&gene=\" + protein_str \n",
    "    print(requestURL)\n",
    "    r = requests.get(requestURL, headers={ \"Accept\" : \"text/x-fasta\"})\n",
    "\n",
    "    if not r.ok:\n",
    "      r.raise_for_status()\n",
    "      sys.exit()\n",
    "\n",
    "    proteome = r.text\n",
    "    return proteome\n",
    "\n",
    "termophilus_taxids = termophilus_taxa[0:19]\n",
    "streptococcus_taxids = streptococcus_taxa[0:19]\n",
    "print(streptococcus_taxids)\n",
    "print(termophilus_taxids)\n",
    "\n",
    "streptococcus_proteome = load_proteome(streptococcus_taxids, -1, protein = [\"LDH\", \"CAS2\", \"CAS3\"])\n",
    "termophilus_proteome = load_proteome(termophilus_taxids, -1, protein = [\"LDH\", \"CAS2\", \"CAS3\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4840\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def proteome2dict(proteome_fasta):\n",
    "    \"\"\"\n",
    "    Returns a dict with keys protein accession and values the list of fasta format for all taxids\n",
    "    This is the basis for clustalw alignments and tree generation\n",
    "    \"\"\"\n",
    "    proteome = {}\n",
    "    key_found = False\n",
    "    for line in proteome_fasta.splitlines():\n",
    "        if len(line) > 0:\n",
    "            if line[0] == \">\":\n",
    "                if key_found:\n",
    "                    if key in proteome:\n",
    "                        proteome[key].append(seq)\n",
    "                    else:\n",
    "                        proteome[key] = [seq]\n",
    "                key_found = True            \n",
    "                search_gene_name = re.search('GN=(\\w*)', line)\n",
    "                key = search_gene_name.group(1).upper()\n",
    "                #print(key)\n",
    "                seq = line + '\\n'\n",
    "            elif key_found:\n",
    "                seq += line + '\\n'\n",
    "    if key_found:\n",
    "        if key in proteome:\n",
    "            proteome[key].append(seq)\n",
    "        else:\n",
    "            proteome[key] = [seq]\n",
    "    return proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'termophilus': {'CAS2': 0.10117692307692308, 'LDH': 0.007116666666666667}, 'streptococcus': {'CAS2': 0.05127011363636366, 'LDHA': 0.09308272727272726, 'LDH_2': 0.1056288888888889, 'LDH_1': 0.12399615384615387, 'LDHA_1': 0.14062904761904763, 'LDHA_3': 0.1900181818181818, 'LDH': 0.038460561797752806, 'LDHA_2': 0.13067238095238096, 'LDHD': 0.04888}}\n"
     ]
    }
   ],
   "source": [
    "# Phylo tree with clustalw. We need to measure the substitution rate.\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "from Bio import Phylo\n",
    "from io import StringIO\n",
    "import os\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "CLUSTALW = r\"./clustalw2\"\n",
    "assert os.path.isfile(CLUSTALW), \"Clustal W executable missing\"\n",
    "plt.rcParams[\"figure.figsize\"] = (20,30)\n",
    "matplotlib.rc('font', size=12)\n",
    "        \n",
    "def compute_mean_subst_rate(proteome, verbose=False, show_tree=False):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    clustalw_cline = ClustalwCommandline(CLUSTALW, infile=proteome + \".fasta\")\n",
    "    stdout, stderr = clustalw_cline()\n",
    "    f = open(proteome + \".dnd\", \"r\")\n",
    "    s_tree = f.read()\n",
    "    f.close()\n",
    "    #print(s_tree)\n",
    "    branch_len = 0\n",
    "    num_branches = 0\n",
    "    search_branch_length = re.findall(':([-.0123456789]*)', s_tree)\n",
    "    for branch_length in search_branch_length:\n",
    "        #print(branch_length)\n",
    "        if branch_length != \"0.00000\":\n",
    "            branch_len += float(branch_length)\n",
    "            num_branches += 1\n",
    "    if verbose: print(branch_len, num_branches, branch_len/num_branches)\n",
    "    if show_tree:\n",
    "        tree = Phylo.read(proteome + \".dnd\", \"newick\")\n",
    "        Phylo.draw(tree)\n",
    "    return branch_len/num_branches\n",
    "\n",
    "def compute_subst_rates(proteome, proteome_name, verbose=False):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    subst_rates = {}\n",
    "    for protein in proteome.keys():\n",
    "        if verbose: print(protein)\n",
    "        protein_sequence = \"\"\n",
    "        # Only for proteins with enough sequences to make a tree\n",
    "        if len(proteome[protein]) >= 3:\n",
    "            for sequence in proteome[protein]:\n",
    "                protein_sequence += sequence\n",
    "            fasta_file_name = proteome_name + \"_\" + protein\n",
    "            f = open(fasta_file_name + \".fasta\", \"w\")\n",
    "            if verbose: print(protein_sequence)\n",
    "            f.write(protein_sequence)\n",
    "            f.close()\n",
    "            mean_subst_rate = compute_mean_subst_rate(fasta_file_name)\n",
    "            subst_rates[protein] = mean_subst_rate\n",
    "    return subst_rates\n",
    "\n",
    "\n",
    "# Proteomes in fasta to dictionnaries\n",
    "proteome_termophilus = proteome2dict(termophilus_proteome)\n",
    "proteome_streptococcus = proteome2dict(streptococcus_proteome)\n",
    "\n",
    "# Compute branch lengths\n",
    "subst_rates_groups = {}\n",
    "subst_rates_groups[\"termophilus\"] = compute_subst_rates(proteome_termophilus, \"termophilus\", False)\n",
    "subst_rates_groups[\"streptococcus\"] = compute_subst_rates(proteome_streptococcus, \"streptococcus\", False)\n",
    "print(subst_rates_groups)\n",
    "\n",
    "# Compute branch ratios\n",
    "# Compute mean of branch ratios and standard deviation\n",
    "# Obtain the most extreme values. These are the proteins that could have been a slowdown or from his initial state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook PY_REST.ipynb to latex\n",
      "[NbConvertApp] Support files will be in PY_REST_files/\n",
      "[NbConvertApp] Making directory PY_REST_files\n",
      "[NbConvertApp] Making directory PY_REST_files\n",
      "[NbConvertApp] Writing 50890 bytes to PY_REST.tex\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#cd /Users/nandoide/Desktop/uni/STRBI.practical\n",
    "jupyter nbconvert --to=latex --template=~/report.tplx PY_REST.ipynb 1> /dev/null\n",
    "pdflatex -shell-escape PY_REST 1> /dev/null"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "nbTranslate": {
   "displayLangs": [
    "en",
    "sp"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "sp",
   "targetLang": "en",
   "useGoogleTranslate": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
