{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Look for coevolution Streptococcus-Lactobacillus\n",
    "\n",
    "Coevolution is everywhere. Because no species can be considered totally isolated from the others. And not only on a biological level, we perceive it in the lines of change in our societies, in political, scientific, religious ideas and even in the evolution of software engineering.\n",
    "\n",
    "The difficulty is to know how much co-evolution is due to the interaction between a pair of species, because the tandem can not be totally isolated from the rest of the universe either. No doubt the problem can not be solved only from the bioinformatic perspective, it is also necessary the contribution of other branches of knowledge such as microbiology, physics, mathematics and computer theory.\n",
    "\n",
    "Our approach consists in comparing the rates of protein evolutionary change between strains of Streptococcus and Lactobacillus linked by symbiotic pathways, and the rates between the rest of species of each genus.\n",
    "\n",
    "We postulate that the proteins that reflects a rate change with more significance, i.e., more speedy or more slowly that intra genus rates, are proteins that could be influenced by the symbiotic environment.\n",
    "\n",
    "To do so we need to compute separately both species of symbiotic tandem. The detailed steps are:\n",
    "\n",
    "1) Obtain all the proteome from this four groups of species:\n",
    "- Genus streptococcus\n",
    "- Strains of streptococcus termophilus.\n",
    "- Genus lactobacillus\n",
    "- Strains of lactobacillus bulgaricus.\n",
    "\n",
    "2) For each protein of each group we obtain the phylogenetic tree and compute the mean branch length. We suppose that we have ultrametricicy or almost ultrametricity. It can be not exact but it could serve as a reference. We use the clulstalw multialignment method, but other methods as T-COFFEE or MUSCLE could be used too. The branch length is a measure of the substitution rate.\n",
    "\n",
    "3) For each protein of the first an second group, we calculate the ration between branch length of termophilus and branch length of protein in genus. And also we compute the overall mean of all ratios. Finally we select the proteins with the most extreme values (80% percentile, two tailed).\n",
    "\n",
    "4) We do the same with lactobacillus (third an fourth groups). At this stage we have two sets of proteins.\n",
    "\n",
    "5) We obtain the biologic pathways of each of the sets of proteins.\n",
    "\n",
    "6) The proteins involved in the same pathways from lactobacillus ans streptoccocus are the target proteins influenced by the coevolution between the two species. It will be necessary a later microbiological study that confirms this expectancies and dive deeply in the details of this coevolution.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtain data from servers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, sys\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "264199\n",
      "Streptococcus thermophilus (strain ATCC BAA-250 / LMG 18311)\n",
      "299768\n",
      "Streptococcus thermophilus (strain CNRZ 1066)\n",
      "322159\n",
      "Streptococcus thermophilus (strain ATCC BAA-491 / LMD-9)\n",
      "767463\n",
      "Streptococcus thermophilus (strain ND03)\n",
      "1042404\n",
      "Streptococcus thermophilus CNCM I-1630\n",
      "1051074\n",
      "Streptococcus thermophilus JIM 8232\n",
      "1073569\n",
      "Streptococcus thermophilus MTCC 5460\n",
      "1073570\n",
      "Streptococcus thermophilus MTCC 5461\n",
      "1091038\n",
      "Streptococcus thermophilus DSM 20617\n",
      "1187956\n",
      "Streptococcus thermophilus MN-ZLW-002\n",
      "1263110\n",
      "Streptococcus thermophilus CAG:236\n",
      "1268061\n",
      "Streptococcus thermophilus DGCC 7710\n",
      "1408178\n",
      "Streptococcus thermophilus ASCC 1275\n",
      "1415776\n",
      "Streptococcus thermophilus TH1435\n",
      "1423145\n",
      "Streptococcus thermophilus TH1436\n",
      "1433288\n",
      "Streptococcus thermophilus MTH17CL396\n",
      "1433289\n",
      "Streptococcus thermophilus M17PTZA496\n",
      "1435972\n",
      "Streptococcus thermophilus TH985\n",
      "1435974\n",
      "Streptococcus thermophilus TH982\n",
      "1435981\n",
      "Streptococcus thermophilus 1F8CT\n",
      "1436725\n",
      "Streptococcus thermophilus TH1477\n",
      "1302\n",
      "Streptococcus gordonii\n",
      "1303\n",
      "Streptococcus oralis\n",
      "1304\n",
      "Streptococcus salivarius\n",
      "1305\n",
      "Streptococcus sanguinis\n",
      "1306\n",
      "Streptococcus sp.\n",
      "1307\n",
      "Streptococcus suis\n",
      "1308\n",
      "Streptococcus thermophilus\n",
      "1309\n",
      "Streptococcus mutans\n",
      "1310\n",
      "Streptococcus sobrinus\n",
      "1311\n",
      "Streptococcus agalactiae\n",
      "1313\n",
      "Streptococcus pneumoniae\n",
      "1314\n",
      "Streptococcus pyogenes\n",
      "1317\n",
      "Streptococcus downei\n",
      "1318\n",
      "Streptococcus parasanguinis\n",
      "1319\n",
      "Streptococcus sp. 'group B'\n",
      "1320\n",
      "Streptococcus sp. group G\n",
      "1324\n",
      "Streptococcus sp. G148\n",
      "1325\n",
      "Streptococcus sp. GX7805\n",
      "1326\n",
      "Streptococcus acidominimus\n",
      "1328\n",
      "Streptococcus anginosus\n",
      "1329\n",
      "Streptococcus canis\n",
      "1332\n",
      "Streptococcus criae\n",
      "1333\n",
      "Streptococcus criceti\n",
      "1334\n",
      "Streptococcus dysgalactiae\n",
      "1335\n",
      "Streptococcus equinus\n",
      "1336\n",
      "Streptococcus equi\n",
      "1337\n",
      "Streptococcus hyointestinalis\n",
      "1338\n",
      "Streptococcus intermedius\n",
      "1339\n",
      "Streptococcus macacae\n",
      "1340\n",
      "Streptococcus porcinus\n",
      "1341\n",
      "Streptococcus ratti\n",
      "1343\n",
      "Streptococcus vestibularis\n",
      "1345\n",
      "Streptococcus ferus\n",
      "1346\n",
      "Streptococcus iniae\n",
      "1348\n",
      "Streptococcus parauberis\n",
      "1349\n",
      "Streptococcus uberis\n",
      "10728\n",
      "Streptococcus pneumoniae phage HB-3\n",
      "10747\n",
      "Streptococcus phage Cp-1\n",
      "10748\n",
      "Streptococcus phage Cp-7\n",
      "10749\n",
      "Streptococcus phage Cp-9\n",
      "12344\n",
      "Streptococcus phage Cp-5\n",
      "12366\n",
      "Streptococcus pyogenes phage H4489A\n",
      "12402\n",
      "Streptococcus phage EJ-1\n",
      "28037\n",
      "Streptococcus mitis\n",
      "29389\n",
      "Streptococcus alactolyticus\n",
      "29390\n",
      "Streptococcus gordonii str. Challis\n",
      "33040\n",
      "Streptococcus milleri\n",
      "33972\n",
      "Streptococcus sp. 'group C'\n",
      "35344\n",
      "Streptococcus pyogenes phage T12\n",
      "36470\n",
      "Streptococcus sp. 'group A'\n",
      "39425\n",
      "Streptococcus phage T270\n",
      "40041\n",
      "Streptococcus equi subsp. zooepidemicus\n",
      "40287\n",
      "Streptococcus mutans serotype C\n",
      "45634\n",
      "Streptococcus cristatus\n",
      "53354\n",
      "Streptococcus gallolyticus\n",
      "55085\n",
      "Streptococcus thoraltensis\n",
      "59310\n",
      "Streptococcus macedonicus\n",
      "64186\n",
      "Streptococcus virus Sfi21\n",
      "68891\n",
      "Streptococcus peroris\n",
      "68892\n",
      "Streptococcus infantis\n",
      "69017\n",
      "Streptococcus sp. (strain 19909)\n",
      "72638\n",
      "Streptococcus virus Sfi19\n",
      "73422\n",
      "Streptococcus phage TP-J34\n",
      "73492\n",
      "Streptococcus pyogenes phage\n",
      "74382\n",
      "Streptococcus phage SFi18\n",
      "76860\n",
      "Streptococcus constellatus\n",
      "78535\n",
      "Streptococcus viridans\n",
      "78541\n",
      "Streptococcus virus Sfi11\n",
      "82269\n",
      "Streptococcus sp. 28D\n",
      "82348\n",
      "Streptococcus pluranimalium\n",
      "82806\n",
      "Streptococcus ovis\n",
      "83545\n",
      "Streptococcus sp. KN1\n",
      "83546\n",
      "Streptococcus sp. KN2\n",
      "83547\n",
      "Streptococcus sp. KN3\n",
      "83549\n",
      "Streptococcus sp. TW1\n",
      "85154\n",
      "Streptococcus phage O1205\n",
      "86065\n",
      "Streptococcus pyogenes phage H10403\n",
      "90410\n",
      "Streptococcus virus DT1\n",
      "99822\n",
      "Streptococcus dysgalactiae subsp. dysgalactiae\n",
      "102143\n",
      "Streptococcus phage S3b\n",
      "102144\n",
      "Streptococcus phage S92\n",
      "102145\n",
      "Streptococcus phage ST3\n",
      "102146\n",
      "Streptococcus phage ST64\n",
      "102147\n",
      "Streptococcus phage Sfi16A\n",
      "102150\n",
      "Streptococcus phage J1\n",
      "102684\n",
      "Streptococcus infantarius\n",
      "102886\n",
      "Streptococcus didelphis\n",
      "104215\n",
      "Streptococcus sp. 1400-98\n",
      "112023\n",
      "Streptococcus virus 7201\n",
      "113107\n",
      "Streptococcus australis\n",
      "114652\n",
      "Streptococcus orisratti\n",
      "116154\n",
      "Streptococcus sp. Z1227\n",
      "116155\n",
      "Streptococcus sp. Z12\n",
      "116156\n",
      "Streptococcus sp. Z89\n",
      "118670\n",
      "Streptococcus sp. PSH2\n",
      "118671\n",
      "Streptococcus sp. PSH1a\n",
      "118672\n",
      "Streptococcus sp. PSH1b\n",
      "119224\n",
      "Streptococcus phocae\n",
      "119602\n",
      "Streptococcus dysgalactiae subsp. equisimilis\n",
      "119603\n",
      "Streptococcus dysgalactiae group\n"
     ]
    }
   ],
   "source": [
    "def load_taxa(scientific_prefix):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    requestURL = \"https://www.ebi.ac.uk/proteins/api/taxonomy/name/\" + scientific_prefix +\\\n",
    "                \"%20?pageNumber=1&pageSize=100&searchType=STARTSWITH&fieldName=SCIENTIFICNAME\"\n",
    "\n",
    "    r = requests.get(requestURL, headers={ \"Accept\" : \"application/json\"})\n",
    "\n",
    "    if not r.ok:\n",
    "      r.raise_for_status()\n",
    "      sys.exit()\n",
    "\n",
    "    jsonBody = json.loads(r.text)\n",
    "    taxa = []\n",
    "    names = []\n",
    "    for taxonomy in jsonBody[\"taxonomies\"]:\n",
    "        print(taxonomy['taxonomyId'])\n",
    "        print(taxonomy['scientificName'])\n",
    "        taxa.append(taxonomy['taxonomyId'])\n",
    "        names.append(taxonomy['scientificName'])\n",
    "    return taxa, names\n",
    "\n",
    "termophilus_taxa,  termophilus_names = load_taxa(\"Streptococcus thermophilus\")\n",
    "streptococcus_taxa,  streptococcus_names = load_taxa(\"Streptococcus\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1302, 1303, 1304, 1305, 1306, 1307, 1308, 1309, 1310, 1311, 1313, 1314, 1317, 1318, 1319, 1320, 1324, 1325, 1326]\n",
      "[264199, 299768, 322159, 767463, 1042404, 1051074, 1073569, 1073570, 1091038, 1187956, 1263110, 1268061, 1408178, 1415776, 1423145, 1433288, 1433289, 1435972, 1435974]\n",
      "1302,1303,1304,1305,1306,1307,1308,1309,1310,1311,1313,1314,1317,1318,1319,1320,1324,1325,1326\n",
      "https://www.ebi.ac.uk/proteins/api/proteins?offset=0&size=-1&taxid=1302,1303,1304,1305,1306,1307,1308,1309,1310,1311,1313,1314,1317,1318,1319,1320,1324,1325,1326&reviewed=false\n",
      "264199,299768,322159,767463,1042404,1051074,1073569,1073570,1091038,1187956,1263110,1268061,1408178,1415776,1423145,1433288,1433289,1435972,1435974\n",
      "https://www.ebi.ac.uk/proteins/api/proteins?offset=0&size=-1&taxid=264199,299768,322159,767463,1042404,1051074,1073569,1073570,1091038,1187956,1263110,1268061,1408178,1415776,1423145,1433288,1433289,1435972,1435974&reviewed=false\n"
     ]
    }
   ],
   "source": [
    "def load_proteome(taxids, size=10, protein=[\"LDH\"]):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    taxids_str = \",\".join(str(x) for x in taxids)\n",
    "    protein_str = \",\".join(x for x in protein)\n",
    "    print(taxids_str)\n",
    "    requestURL = \"https://www.ebi.ac.uk/proteins/api/proteins?offset=0&size=\" + str(size) + \"&taxid=\" +\\\n",
    "                    taxids_str + \"&reviewed=false\"\n",
    "    if protein != []:\n",
    "        requestURL += \"&gene=\" + protein_str \n",
    "    print(requestURL)\n",
    "    r = requests.get(requestURL, headers={ \"Accept\" : \"text/x-fasta\"})\n",
    "\n",
    "    if not r.ok:\n",
    "      r.raise_for_status()\n",
    "      sys.exit()\n",
    "\n",
    "    proteome = r.text\n",
    "    return proteome\n",
    "\n",
    "# We limit to 20 taxids (it's the maximum for webservice and it should be enough\n",
    "termophilus_taxids = termophilus_taxa[0:19]\n",
    "streptococcus_taxids = streptococcus_taxa[0:19]\n",
    "print(streptococcus_taxids)\n",
    "print(termophilus_taxids)\n",
    "\n",
    "#streptococcus_proteome = load_proteome(streptococcus_taxids, -1, protein = [\"LDH\", \"CAS2\", \"CAS3\"])\n",
    "#termophilus_proteome = load_proteome(termophilus_taxids, -1, protein = [\"LDH\", \"CAS2\", \"CAS3\"])\n",
    "\n",
    "streptococcus_proteome_complete = load_proteome(streptococcus_taxids, -1, [])\n",
    "termophilus_proteome_complete = load_proteome(termophilus_taxids, -1, [])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "def dump(obj, obj_name):\n",
    "    binary_file = open(obj_name + '.bin',mode='wb')\n",
    "    pickled_obj = pickle.dump(obj, binary_file)\n",
    "    binary_file.close()\n",
    "\n",
    "def load(obj_name):\n",
    "    with (open(obj_name + \".bin\", \"rb\")) as openfile:\n",
    "        while True:\n",
    "            try:\n",
    "                obj = pickle.load(openfile)\n",
    "            except EOFError:\n",
    "                break\n",
    "    return obj\n",
    "\n",
    "dump(streptococcus_proteome_complete, \"streptococcus_proteome_complete\")\n",
    "dump(termophilus_proteome_complete, \"termophilus_proteome_complete\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">tr|V6CG59|V6CG59_STRTN Proteolysis tag peptide encoded by tmRNA Strep_therm_CNRZ10 (Fragment) OS=Streptococcus thermophilus (strain ND03) OX=767463 GN=tmRNA Strep_therm_CNRZ10 PE=4 SV=1\n",
      "AKNTNSYAVAA\n",
      ">tr|V6CG54|V6CG54_STRTR Proteolysis tag peptide encoded by tmRNA Strep_therm (Fragment) OS=Streptococcus thermophilus MN-ZLW-002 OX=1187956 GN=tmRNA Strep_therm PE=4 SV=1\n",
      "AKNTNSYAVAA\n",
      ">tr|V6CG50|V6CG50_STRTR Proteolysis tag peptide encoded by tmRNA Strep_therm_CNRZ10 (Fragment) OS=Streptococcus thermophilus MTCC 5461 OX=1073570 GN=tmRNA Strep_therm_CNRZ10 PE=4 SV=1\n",
      "AKNTNSYAVAA\n",
      ">tr|V6BJW4|V6BJW4_STRTR Proteolysis tag peptide encoded by tmRNA Strep_therm_CNRZ10 (Fragment) OS=Streptococcus thermophilus MTCC 5460 OX=1073569 GN=tmRNA Strep_therm_CNRZ10 PE=4 SV=1\n",
      "AKNTNSYAVAA\n",
      ">tr|V6CE29|V6CE29_STRT1 Proteolysis tag peptide encoded by tmRNA Strep_therm_CNRZ10 (Fragment) OS=Streptococcus thermophilus (strain CNRZ 1066) OX=299768 GN=tmRNA Strep_therm_CNRZ10 PE=4 SV=1\n",
      "AKNTNSYAVAA\n",
      ">tr|A0A1L1QK15|A0A1L1Q\n",
      ">tr|D1MQ05|D1MQ05_STRDO Sortase OS=Streptococcus downei OX=1317 GN=srtA PE=4 SV=1\n",
      "MKRTSSRNSKGRKEKESSKKKHHWLRTILVLFLMVVGLALVFNRSIRNSVIAWNTNKYQV\n",
      "SKVDKKTLTKNKKAKVNYDFDSVKSISAQSVISSQMDAQDLPVIGGIAIPDLELNLPIFK\n",
      "GLGNTELSYGAGTMKESQVMGGENNYALASHHVFGVNGASKMLFSPLENAKNGMEIYLTD\n",
      "KNKVYTYIISEVKTVEPSDVAVINDTPGAKNLTLVTCDDAEATNRIIVSANYKEEVSYDK\n",
      "ASQKMIEAFNRSYNQLSL\n",
      ">tr|A7VLR2|A7VLR2_STRDO Autolysin OS=Streptococcus downei OX=1317 GN=atlh PE=4 SV=1\n",
      "MRKKMYKVKKHWVIASAAALTVLGTSGLVAADEQSYPQTGASRELASVSTDPSVTATDTQ\n",
      "DSPKSVENTSQPASTSTLDDSSQEGQNQLVDETSQSSQSEPAGSETISDWETDNQSASEV\n",
      "QSSQRDSEAQIQAPASTNQVDPSDEEDQLQVADQTINEGSRVLRASAGATVARATAAGQS\n",
      "SSEVTSASVSGKTLIIHYNRSIASNEALYFAVWTEENDQDDMVWYQANAQGAAYIDLSKH\n",
      "RFYGKYIIHTYSVINGQNIGRNATSIMVNPPKISSKIEATSNGDYKLTVGNVSSDITEVF\n",
      "VPVWSDKNDQDDMIWYHADKVNANTYSLTIKPANHHNDFDHYSVHIYGQSSITRGLIGLS\n",
      "ATPGFERKETASSSKVAPQNKVRASLGANGIQLHLDTNEAADLSNIYFAVWSQVGDQDDI\n",
      "HWYPADGSLSALAPYSNHSGYGTYHIHTYANKNSKFVGINTTSIEVPSPSASAKISKQDD\n",
      "ATYNVKIDKVSSYITSVLVPVWTQKNDQDDIVWYPASKQSDGSYSATIKLKDHNFESGKY\n",
      "LVHLYGYSSLEGGRLVGLGATD\n"
     ]
    }
   ],
   "source": [
    "print(termophilus_proteome_complete[0:1000])\n",
    "print(streptococcus_proteome_complete[0:1000])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute substitution rates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def proteome2dict(proteome_fasta):\n",
    "    \"\"\"\n",
    "    Returns a dict with keys protein accession and values the list of fasta format for all taxids\n",
    "    This is the basis for clustalw alignments and tree generation\n",
    "    \"\"\"\n",
    "    proteome = {}\n",
    "    key_found = False\n",
    "    for line in proteome_fasta.splitlines():\n",
    "        if len(line) > 0:\n",
    "            if line[0] == \">\":\n",
    "                if key_found:\n",
    "                    if key in proteome:\n",
    "                        proteome[key].append(seq)\n",
    "                    else:\n",
    "                        proteome[key] = [seq]  \n",
    "                search_gene_name = re.search('GN=(\\w*)', line)\n",
    "                if search_gene_name:\n",
    "                    key = search_gene_name.group(1).upper()\n",
    "                    key_found = True  \n",
    "                #print(key)\n",
    "                seq = line + '\\n'\n",
    "            elif key_found:\n",
    "                seq += line + '\\n'\n",
    "    if key_found:\n",
    "        if key in proteome:\n",
    "            proteome[key].append(seq)\n",
    "        else:\n",
    "            proteome[key] = [seq]\n",
    "    return proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Phylo tree with clustalw. We need to measure the substitution rate.\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "from Bio import Phylo\n",
    "from io import StringIO\n",
    "import os\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "CLUSTALW = r\"./clustalw2\"\n",
    "assert os.path.isfile(CLUSTALW), \"Clustal W executable missing\"\n",
    "plt.rcParams[\"figure.figsize\"] = (20,30)\n",
    "matplotlib.rc('font', size=12)\n",
    "        \n",
    "def compute_mean_subst_rate(proteome, verbose=False, show_tree=False):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    clustalw_cline = ClustalwCommandline(CLUSTALW, infile=proteome + \".fasta\")\n",
    "    stdout, stderr = clustalw_cline()\n",
    "    f = open(proteome + \".dnd\", \"r\")\n",
    "    s_tree = f.read()\n",
    "    f.close()\n",
    "    #print(s_tree)\n",
    "    branch_len = 0\n",
    "    num_branches = 0\n",
    "    search_branch_length = re.findall(':([-.0123456789]*)', s_tree)\n",
    "    for branch_length in search_branch_length:\n",
    "        #print(branch_length)\n",
    "        if branch_length != \"0.00000\":\n",
    "            branch_len += float(branch_length)\n",
    "            num_branches += 1\n",
    "    if num_branches > 0:\n",
    "        rate = branch_len/num_branches\n",
    "    else:\n",
    "        rate = -1\n",
    "    if verbose: print(branch_len, num_branches, rate)\n",
    "    if show_tree:\n",
    "        tree = Phylo.read(proteome + \".dnd\", \"newick\")\n",
    "        Phylo.draw(tree)\n",
    "    return rate\n",
    "\n",
    "def compute_subst_rates(proteome, proteome_keys, proteome_name, verbose=False):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    subst_rates = {}\n",
    "    for protein in proteome_keys:\n",
    "        if verbose: print(protein)\n",
    "        protein_sequence = \"\"\n",
    "        # Only for proteins with enough sequences to make a tree\n",
    "        if len(proteome[protein]) >= 3:\n",
    "            for sequence in proteome[protein]:\n",
    "                protein_sequence += sequence\n",
    "            fasta_file_name = proteome_name + \"_\" + protein\n",
    "            f = open(fasta_file_name + \".fasta\", \"w\")\n",
    "            if verbose: print(protein_sequence)\n",
    "            f.write(protein_sequence)\n",
    "            f.close()\n",
    "            mean_subst_rate = compute_mean_subst_rate(fasta_file_name)\n",
    "            subst_rates[protein] = mean_subst_rate\n",
    "    return subst_rates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fasta to dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TMRNA', 'BN551_00358', 'GALT', 'LACZ', 'SBCD', 'GALR', 'DNAN', 'STU0007', 'FTSH', 'MREC', 'ARAT', 'PURL', 'STU0044', 'STU0052', 'STU0053', 'TAG', 'STU0075', 'STU0082', 'LABC', 'STU0110', 'STU0113', 'ILVA', 'STU0161', 'STU0182', 'STU0202', 'STU0208', 'STU0251', 'STU0258', 'STU0267', 'CBIM', 'STU0297', 'TRKH2', 'DNAB', 'SERS', 'STU0330', 'STU0334', 'STU0338', 'STU0339', 'METB1', 'STU0358', 'LIVM', 'HPF', 'FABF', 'FRUR', 'STU0422', 'STU0435', 'PHOH', 'STU0448', 'METS', 'STU0452', 'PRTM', 'ARGB', 'STU0468', 'STU0473', 'STU0474', 'STU0475', 'FTSW', 'DNAH', 'NAGA', 'STU0510', 'STU0516', 'STU0539', 'STU0551', 'STU0557', 'STU0565', 'HLYIII', 'STU0580', 'STU0595', 'MURN', 'NNRD', 'RNR', 'STU0631', 'STU0636', 'AROE', 'MIP', 'CAS2', 'STU0665', 'STU0668', 'STU0672', 'STU0675', 'STU0678', 'STU0679', 'STU0681', 'STU0693', 'STU0695', 'GPMC', 'STU0702', 'STU0704', 'LEMA', 'STU0721', 'APBE', 'DLTX', 'STU0811', 'STU0819', 'STU0829', 'ADCA', 'MUR2', 'STU0876', 'STU0877', 'STHIM']\n",
      "['SRTA', 'ATLH', 'DBLB', 'RPOA', 'PHES', 'DEX', 'RPOB', 'GBPC3', 'GROEL', 'GTF', 'GROS', 'SODA', 'DBLA', 'GYRB', 'RECA', 'ATPA', 'ERMTR', 'MEF', 'EMM', 'TUF', 'PAH', 'GBPC2', 'GBPC1', 'ERMB', 'GFBA', 'SILB', 'SILCR', 'NADE', 'ENO1', 'SPAA', 'PROTEIN', 'EMML', 'BLPM', 'SILD', 'SILC', 'SILA', 'SSO1', 'GTFU', 'GLUCOSYLTRANSFERASE', 'DEI', 'PAR', 'CAS1', 'SILE', 'SU1', 'SU2', 'GTFI', 'GROES', 'DBLC', 'CPN60', 'CAS2', 'FTSK', 'GTFS', 'CMK', 'GALU', 'SSO2', 'LCD', 'GTFT', 'PAG', 'FOLP', 'RODA', 'HISC', 'ATLG', 'GTFC', 'DRPA', 'MUTF', 'SRLB', 'SRLE', 'SRLM', 'MUTC', 'ORF', 'SLOB', 'MUTA', 'MUTP', 'MUTG', 'SATC', 'MUTR', 'COMC', 'RMPB', 'PSAA', 'MEPA', 'MBRA', 'RGPE', 'RGPA', 'RGPF', 'SMBG', 'SMBB', 'GBPC', 'A6J86_007565', 'LPP', 'DDL', 'DLTD', 'ORF2', 'RGPBC', 'GTFB', 'GTFD', 'PYK', 'SPAP', 'PPAC', 'LEPC', 'ATLA']\n"
     ]
    }
   ],
   "source": [
    "#Proteomes in fasta to dictionaries\n",
    "RESTART = True\n",
    "VERBOSE = True\n",
    "\n",
    "if RESTART:\n",
    "    termophilus_proteome_fasta = load(\"termophilus_proteome_complete\")\n",
    "    streptococcus_proteome_fasta = load(\"streptococcus_proteome_complete\")\n",
    "\n",
    "#print(termophilus_proteome_fasta)\n",
    "proteome_termophilus = proteome2dict(termophilus_proteome_fasta)\n",
    "proteome_streptococcus = proteome2dict(streptococcus_proteome_fasta)\n",
    "\n",
    "dump(proteome_termophilus, \"proteome_termophilus\")\n",
    "dump(proteome_streptococcus, \"proteome_streptococcus\")\n",
    "\n",
    "if VERBOSE: print(list(proteome_termophilus.keys())[0:100])\n",
    "if VERBOSE: print(list(proteome_streptococcus.keys())[0:100])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Substitution rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TMRNA', 'GALT', 'LACZ', 'SBCD', 'GALR', 'DNAN', 'FTSH', 'MREC', 'ARAT', 'PURL', 'TAG', 'LABC', 'ILVA', 'CBIM', 'TRKH2', 'DNAB', 'SERS', 'LIVM', 'HPF', 'FABF']\n",
      "Proteins we need to process: 964\n",
      "Proteins we want to process: 20\n",
      "{'termophilus': {'TMRNA': -1, 'GALT': 0.014773333333333333, 'LACZ': 0.00292, 'SBCD': 0.00245625, 'FTSH': 0.00262, 'ILVA': 0.0024, 'SERS': 0.001318888888888889, 'HPF': 0.01282, 'FABF': 0.31672}, 'streptococcus': {'TMRNA': 0.10939888888888888, 'GALT': 0.011721821086261982, 'LACZ': 0.028357828947368432, 'SBCD': 0.05702195652173911, 'GALR': 0.0795148, 'DNAN': 0.01873888888888889, 'FTSH': 0.017297005347593573, 'MREC': 0.06296962962962963, 'ARAT': 0.17196846153846151, 'PURL': 0.012450666666666667, 'TAG': 0.02267730769230769, 'ILVA': 0.012551562499999992, 'CBIM': 0.10676333333333335, 'TRKH2': 0.16780555555555557, 'DNAB': 0.09844512820512821, 'SERS': 0.008190401785714287, 'LIVM': 0.027198333333333328, 'HPF': 0.03110622641509432, 'FABF': 0.029105231788079465}}\n"
     ]
    }
   ],
   "source": [
    "#Proteomes in fasta to dictionaries\n",
    "RESTART = True\n",
    "VERBOSE = True\n",
    "#Protein limit\n",
    "LIMIT = 20 \n",
    "\n",
    "if RESTART:\n",
    "    proteome_termophilus = load(\"proteome_termophilus\")\n",
    "    proteome_streptococcus = load(\"proteome_streptococcus\")\n",
    "\n",
    "# Compute keys to process: only the keys that are included in both groups\n",
    "proteome_keys = []\n",
    "for key in proteome_termophilus.keys():\n",
    "    if key in proteome_streptococcus.keys():\n",
    "        proteome_keys.append(key)\n",
    "\n",
    "if VERBOSE: print(proteome_keys[0:LIMIT])\n",
    "print(\"Proteins we need to process:\", len(proteome_keys))\n",
    "limit =  min(len(proteome_keys), LIMIT) \n",
    "print(\"Proteins we want to process:\", limit)\n",
    "\n",
    "# Compute branch lengths\n",
    "subst_rates_groups = {}\n",
    "subst_rates_groups[\"termophilus\"] = compute_subst_rates(proteome_termophilus, proteome_keys[0:limit],\n",
    "                                                        \"termophilus\", False)\n",
    "subst_rates_groups[\"streptococcus\"] = compute_subst_rates(proteome_streptococcus, proteome_keys[0:limit], \n",
    "                                                          \"streptococcus\", False)\n",
    "\n",
    "dump(subst_rates_groups, \"subst_rates_groups\")\n",
    "if VERBOSE: print(subst_rates_groups)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selection of first set of target proteins\n",
    "At this step is necessary to calculate the ratios, the means of the ratios and select the more extreme between them (two tailed percentile 80%)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute branch ratios\n",
    "# Compute mean of branch ratios and standard deviation\n",
    "# Obtain the most extreme values. \n",
    "# These are the proteins that could have been a slowdown or from his initial state\n",
    "import statistics as stats\n",
    "\n",
    "def compute_branch_ratios(subst_rates_groups, group1, group2, verbose=False):\n",
    "    \"\"\" \n",
    "    For every protein in group1 that has counterpart in group2 calculate the ratio of\n",
    "    branch lengths.\n",
    "    \n",
    "    Returns:\n",
    "        dict of string, float: ratios by protein\n",
    "    \"\"\"\n",
    "    ratios = {}\n",
    "    for protein in subst_rates_groups[group1].keys():\n",
    "        if protein in subst_rates_groups[group2].keys():\n",
    "            ratio = subst_rates_groups[group1][protein]/(subst_rates_groups[group2][protein] + 0.00001)\n",
    "            if subst_rates_groups[group1][protein] != -1 and subst_rates_groups[group2][protein] != -1:\n",
    "                ratios[protein] = ratio\n",
    "            if verbose: print(group1, protein, \n",
    "                              subst_rates_groups[group1][protein], \n",
    "                              subst_rates_groups[group2][protein],\n",
    "                              ratio)\n",
    "    return ratios\n",
    "\n",
    "def compute_mean_std(ratios, verbose=False):\n",
    "    \"\"\"\n",
    "    Calculate mean and std for ratios\n",
    "    \"\"\"\n",
    "    ratios_list =  list(ratios.values())\n",
    "    if verbose: print(\"ratios_list\",ratios_list)\n",
    "    ratios_mean = stats.mean(ratios_list)\n",
    "    ratios_stdev = stats.stdev(ratios_list)\n",
    "    return ratios_mean, ratios_stdev\n",
    "\n",
    "def filter_target_proteins(ratios, mean, std, n_std, verbose=False):\n",
    "    \"\"\"\n",
    "    Filter proteins that are n_std > mean or n_std < mean\n",
    "    \"\"\"\n",
    "    proteins = []\n",
    "    for protein in ratios.keys():\n",
    "        if verbose: print(protein, ratios[protein], mean, std)\n",
    "        if ratios[protein] > mean + n_std * std or ratios[protein] < mean - n_std * std:\n",
    "            proteins.append(protein)\n",
    "    return proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selection of target proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Target proteins: ['FABF']\n"
     ]
    }
   ],
   "source": [
    "RESTART = True\n",
    "VERBOSE = False\n",
    "\n",
    "if RESTART:\n",
    "    subst_rates_groups = load(\"subst_rates_groups\")\n",
    "\n",
    "rat = compute_branch_ratios(subst_rates_groups, \"termophilus\", \"streptococcus\", VERBOSE)   \n",
    "if VERBOSE: print(rat)\n",
    "\n",
    "mean, std = compute_mean_std(rat, VERBOSE)\n",
    "if VERBOSE: print(\"\\nMean\", mean, \"Standard deviation\", std, \"\\n\")\n",
    "\n",
    "prot = filter_target_proteins(rat, mean, std, 0.8, VERBOSE)\n",
    "\n",
    "print(\"Target proteins:\", prot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtain protein pathways from servers.\n",
    "At this stage we obtain the pathways for proteins."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selection of final set of target proteins\n",
    "At this stage we select the proteins that belong tho the same pathways on Lactobacillus and Streptoccus. These are the proteins that could be affected by co-evolucion."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate document outputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook coevolution.ipynb to latex\n",
      "[NbConvertApp] Writing 41101 bytes to coevolution.tex\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#cd /Users/nandoide/Desktop/uni/STRBI.practical\n",
    "jupyter nbconvert --to=latex --template=~/report.tplx coevolution.ipynb 1> /dev/null\n",
    "pdflatex -shell-escape coevolution 1> /dev/null"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "latex_metadata": {
   "author": "Elena Montenegro, Sara González, Sofía Cerdá, Fernando Freire",
   "title": "Look for co-evolution"
  },
  "nbTranslate": {
   "displayLangs": [
    "en",
    "sp"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "auto",
   "targetLang": "en",
   "useGoogleTranslate": true
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
